src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_StringConcat.cs

```diff
@@ -299,33 +334,208 @@ private BoundExpression RewriteStringConcatenationOneExpr(BoundExpression lowere
+
+                var concatCall = BoundCall.Synthesized(syntax, receiverOpt: null, initialBindingReceiverIsSubjectToCloning: ThreeState.Unknown, stringConcatSpans, wrappedLeft, wrappedRight);
+
+                return new BoundSequence(
+                    syntax,
+                    locals.ToImmutableAndFree(),
+                    ImmutableArray<BoundExpression>.Empty,
+                    concatCall,
+                    concatCall.Type);
+            }
+
+            Debug.Assert(!leftType.IsReadOnlySpanChar() && !rightType.IsReadOnlySpanChar());
+
+            if (leftIsChar)
+            {
+                loweredLeft = WrapCharExprIntoToStringCall(loweredLeft);
+            }
+
+            if (rightIsChar)
+            {
+                loweredRight = WrapCharExprIntoToStringCall(loweredRight);
+            }
 
             var method = UnsafeGetSpecialTypeMethod(syntax, SpecialMember.System_String__ConcatStringString);
             Debug.Assert((object)method != null);
 
             return BoundCall.Synthesized(syntax, receiverOpt: null, initialBindingReceiverIsSubjectToCloning: ThreeState.Unknown, method, loweredLeft, loweredRight);
         }
 
-        private BoundExpression RewriteStringConcatenationThreeExprs(SyntaxNode syntax, BoundExpression loweredFirst, BoundExpression loweredSecond, BoundExpression loweredThird)
+        private BoundExpression RewriteStringConcatenationThreeExprs(SyntaxNode syntax, BoundExpression loweredFirst, BoundExpression loweredSecond, BoundExpression loweredThird, ArrayBuilder<LocalSymbol> previousLocals)
         {
-            Debug.Assert(loweredFirst.HasAnyErrors || loweredFirst.Type is { } && loweredFirst.Type.IsStringType());
-            Debug.Assert(loweredSecond.HasAnyErrors || loweredSecond.Type is { } && loweredSecond.Type.IsStringType());
-            Debug.Assert(loweredThird.HasAnyErrors || loweredThird.Type is { } && loweredThird.Type.IsStringType());
+            var firstType = loweredFirst.Type;
+            var secondType = loweredSecond.Type;
+            var thirdType = loweredThird.Type;
+
+            Debug.Assert(loweredFirst.HasAnyErrors || firstType is not null && (firstType.IsStringType() || firstType.IsCharType() || firstType.IsReadOnlySpanChar()));
+            Debug.Assert(loweredSecond.HasAnyErrors || secondType is not null && (secondType.IsStringType() || secondType.IsCharType() || secondType.IsReadOnlySpanChar()));
+            Debug.Assert(loweredThird.HasAnyErrors || thirdType is not null && (thirdType.IsStringType() || thirdType.IsCharType() || thirdType.IsReadOnlySpanChar()));
+
+            var firstIsChar = firstType?.IsCharType() == true;
+            var secondIsChar = secondType?.IsCharType() == true;
+            var thirdIsChar = thirdType?.IsCharType() == true;
+
+            if ((firstIsChar || secondIsChar || thirdIsChar || firstType.IsReadOnlySpanChar() || secondType.IsReadOnlySpanChar() || thirdType.IsReadOnlySpanChar()) &&
```

Comment:
>firstIsChar || secondIsChar || thirdIsChar [](http://example.com/codeflow?start=17&length=42)

Could you point me to unit-tests for which
1. ```firstIsChar``` is true
2.  ```secondChar``` is true
3.  ```thirdIsChar``` is true

And same for the span types.